src/stan/callbacks/json_writer.hpp

```diff
@@ -0,0 +1,375 @@
+      new_value[pos + count] = chars_to_replace[idx][0];
+      new_value[pos + count + 1] = chars_to_replace[idx][1];
+      pos += 1;
+      count++;
+      prev_pos = pos;
+    }
+    for (int i = prev_pos; i < value.size(); ++i) {
+      new_value[i + count] = value[i];
+    }
+    // Shrink any unused space
+    new_value.resize(value.size() + count);
+    return new_value;
+  }
+
+  /**
+   * Writes key plus colon for key-value pair.
+   *
+   * @param[in] key member name.
+   */
+  void write_key(const std::string& key) { *output_ << "\"" << key << "\" : "; }
+  /**
+   * Writes a set of comma separated strings.
+   * Strings are cleaned to escape special characters.
+   *
+   * @param[in] v Values in a std::vector
+   */
+  void write_vector(const std::vector<std::string>& v) {
+    if (v.empty()) {
+      return;
+    }
+    *output_ << "[ ";
+    auto last = v.end();
+    --last;
+    for (auto it = v.begin(); it != last; ++it) {
+      *output_ << process_string(*it) << ", ";
+    }
+    *output_ << v.back() << " ]";
+  }
+
+  /**
+   * Writes a set of comma separated values.
+   *
+   * @param[in] v Values in a std::vector
+   */
+  template <class T>
+  void write_vector(const std::vector<T>& v) {
+    if (v.empty()) {
+      return;
+    }
+    *output_ << "[ ";
+    auto last = v.end();
+    --last;
+    for (auto it = v.begin(); it != last; ++it) {
+      *output_ << *it << ", ";
+    }
+    *output_ << v.back() << " ]";
+  }
+
+ public:
+  json_writer() : output_(nullptr) {}
+  /**
+   * Constructs a json writer with an output stream.
+   *
+   * @param[in, out] output unique pointer to a type inheriting from
+   * `std::ostream`
+   */
+  explicit json_writer(std::unique_ptr<Stream, Deleter>&& output)
+      : output_(std::move(output)) {
+    if (output_ == nullptr)
+      throw std::invalid_argument("writer cannot be null");
+  }
+
+  json_writer(json_writer& other) = delete;
+  json_writer(json_writer&& other) : output_(std::move(other.output_)) {}
+
+  ~json_writer() {}
+
+  /**
+   * Writes "{", initial token of a JSON record.
+   */
+  void begin_record() {
+    write_record_comma_if_needed();
+    *output_ << "{";
+    record_needs_comma_ = false;
+    record_depth_++;
+  }
+
+  /**
+   * Writes "\"key\" : {", initial token of a named JSON record.
+   * @param[in] key The name of the record.
+   */
+  void begin_record(const std::string& key) {
+    write_record_comma_if_needed();
+    *output_ << "\"" << key << "\": {";
+    record_needs_comma_ = false;
+    record_depth_++;
+  }
+  /**
+   * Writes "}", final token of a JSON record.
+   */
+  void end_record() {
+    if (record_depth_ > 0) {
+      *output_ << "}";
+      record_depth_--;
+      if (record_depth_ > 0) {
+        record_needs_comma_ = true;
+      }
+      record_internal_needs_comma_ = false;
+    } else {
+      throw std::runtime_error(
+          "Attempted to end record, but there is no open record.");
+    }
```

Comment:
We throw an error if you try to close a record which doesn't exist, which I think is great. However, we don't do any other checking that json is valid, e.g. that we are inside an object before any key is written.

I'd like to have that check such that it is a runtime error to do 

```c++
json_writer out(...);
out.write("hello", "world");
```
which would produce `hello: "world"`, while the nearest valid JSON is `{ hello: "world" }`

Reply:
idt there is any way to meaningfully validate a steam of json that is not also very difficult (hence why it's so hard / weird to do streaming json in rapidjson.)

Take the following case as an example where the try throws and someone wants to attempt to fix the output file

```c++
std::string write_to_file = "./path/to/my/file.json";
try {
  json_writer out(write_to_file);
  out.begin_record();
  out.write("hello", "world");
  // Other stuff happens that could throw
  out.end_record();
} catch (const std::exception& e) {
  // checks if a simple fix like adding double quote or a comma can produce a valid line of json
  if (attempt_to_fix_json(write_to_file)) {
    json_writer out(write_to_file);
    out.write("error",  e.what());
    out.end_record();
  }   
  // If the json is now invalid we stop, else we can keep going
  if (invalid_json(write_to_file)) {
      throw e;
  }
}
```

in the above calling `end_record()` would be a valid attempt to fix the file, but we've the added state from checking if `begin_record()` is called already would cause this to fail

Reply:
I think that the best thing to do is remove the `end_record` check and document that this is a streaming writer which doesn't perform validation on the JSON, because, as Steve points out, we don't know anything about the output stream we're writing to.

Reply:
updated initial class doc comment, removed check on balanced close objects.